src/interrupt.c

```diff
@@ -34,119 +34,108 @@ static const char * exception_names[] = {
-			console_printf("interrupt: illegal page access at vaddr %x\n",vaddr);
-			process_dump(current);
-			process_exit(0);
-		} else {
-			printf("interrupt: page fault at %x\n",vaddr);
-			printf("please write a fault handler in interrupt.c!\n");
-			printf("kernel halted.\n");
-			halt();
-		}
-	} else {
-		console_printf("interrupt: exception %d: %s (code %x)\n",i,exception_names[i],code);
-		process_dump(current);
-	}
-
-	if(current) {
-		process_exit(0);
-	} else {
-		console_printf("interrupt: exception in kernel code!\n");
-		halt();
-	}
+static void unknown_exception(int i, int code) {
+    unsigned vaddr, paddr;
+
+    if (i==14) {
+        asm("mov %%cr2, %0" : "=r" (vaddr));
+        if (pagetable_getmap(current->pagetable, vaddr, &paddr)) {
+            console_printf("interrupt: illegal page access at vaddr %x\n", vaddr);
+            process_dump(current);
+            process_exit(0);
+        } else {
+            printf("interrupt: page fault at %x\n", vaddr);
+            printf("please write a fault handler in interrupt.c!\n");
+            printf("kernel halted.\n");
+            halt();
+        }
+    } else {
+        console_printf("interrupt: exception %d: %s (code %x)\n", i, exception_names[i], code);
+        process_dump(current);
+    }
+
+    if (current) {
+        process_exit(0);
+    } else {
+        console_printf("interrupt: exception in kernel code!\n");
+        halt();
+    }
 }
 
-static void unknown_hardware( int i, int code )
-{
-	if(!interrupt_spurious[i]) {
-		console_printf("interrupt: spurious interrupt %d\n",i);
-	}
-	interrupt_spurious[i]++;
+static void unknown_hardware(int i, int code) {
+    if (!interrupt_spurious[i]) {
+        console_printf("interrupt: spurious interrupt %d\n", i);
+    }
+    interrupt_spurious[i]++;
 }
 
-void interrupt_register( int i, interrupt_handler_t handler )
-{
-	interrupt_handler_table[i] = handler;
+void interrupt_register(int i, interrupt_handler_t handler) {
+    interrupt_handler_table[i] = handler;
 }
 
-static void interrupt_acknowledge( int i )
-{
-	if(i<32) {
-		/* do nothing */
-	} else {
-		pic_acknowledge(i-32);
-	}
+static void interrupt_acknowledge(int i) {
+    if (i<32) {
+        /* do nothing */
+    } else {
+        pic_acknowledge(i-32);
+    }
 }
 
-void interrupt_init()
-{
-	int i;
-	pic_init(32,40);
-	for(i=32;i<48;i++) {
-		interrupt_disable(i);
-		interrupt_acknowledge(i);
-	}
-	for(i=0;i<32;i++) {
-		interrupt_handler_table[i] = unknown_exception;
-		interrupt_spurious[i] = 0;
-		interrupt_count[i] = 0;
-	}
-	for(i=32;i<48;i++) {
-		interrupt_handler_table[i] = unknown_hardware;
-		interrupt_spurious[i] = 0;
-		interrupt_count[i] = 0;
-	}
-
-	interrupt_unblock();
-
-	console_printf("interrupt: ready\n");
+void interrupt_init() {
+    int i;
+    pic_init(32, 40);
+    for (i=32; i<48; i++) {
+        interrupt_disable(i);
+        interrupt_acknowledge(i);
+    }
+    for (i=0; i<32; i++) {
```

Comment:
The only reasonable way to do spaces between operators is to use `indent`. I've been playing with indent, and I can emulate most of our style guide, though it adds spaces before the first element and after the last element of lists: \{ 1, 2, 3 }`. I can push up my indent file and you can see how you like it.